libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -26,21 +26,68 @@
 #include <libyul/optimiser/Semantics.h>
 #include <libyul/Exceptions.h>
 #include <libyul/AsmData.h>
+#include <libyul/backends/evm/EVMDialect.h>
 
 #include <libdevcore/CommonData.h>
 
 #include <boost/range/adaptor/reversed.hpp>
+#include <boost/range/algorithm_ext/erase.hpp>
 
 using namespace std;
 using namespace dev;
 using namespace yul;
 
+
+// For elementary statements, we check if it is an SSTORE(x, y)
+//   If yes, visit the statement. Then record that fact and clear all storage slots t
+//     where we cannot prove x != t or y == m_storage[t] using the current values of the variables x and t.
+//   Otherwise, determine if the statement invalidates storage. If yes, clear all knowledge
+//     about storage before visiting the statement. Then visit the statement.
+//
+// For forward-joining control flow, storage information from the branches is combined.
+// If the keys or values are different or non-existent in one branch, the key is deleted.
+
+void DataFlowAnalyzer::operator()(ExpressionStatement& _statement)
+{
+	if (auto vars = isSimpleStore(dev::eth::Instruction::SSTORE, _statement))
+	{
+		ASTModifier::operator()(_statement);
+		m_storage.set(vars->first, vars->second);
+		set<YulString> keysToErase;
+		for (auto const& item: m_storage.values)
+			if (!(
+				m_knowledgeBase.knownToBeDifferent(vars->first, item.first) ||
+				m_knowledgeBase.knownToBeEqual(vars->second, item.second)
+			))
+				keysToErase.insert(item.first);
+		for (YulString const& key: keysToErase)
+			m_storage.eraseKey(key);
+	}
+	else if (auto vars = isSimpleStore(dev::eth::Instruction::MSTORE, _statement))
+	{
+		ASTModifier::operator()(_statement);
+		m_memory.set(vars->first, vars->second);
+		set<YulString> keysToErase;
+		for (auto const& item: m_storage.values)
+			if (!m_knowledgeBase.knownToBeDifferentByAtLeast32(vars->first, item.first))
```

Comment:
We can't check for equal values here?

Reply:
no, because they addresses overlap:
```
mstore(0, 1)
mstore(1, 1)
```
the second `mstore` still changes the value of `mload(0)`.